vault/eventbus/filter.go

```diff
@@ -50,71 +69,291 @@ func (nf *NodeFilter) match(ns *namespace.Namespace, eventType logical.EventType
+	// fast check when we don't need to create the Cond
+	f.lock.RLock()
+	n, ok := f.notifyChanges[c]
+	f.lock.RUnlock()
+	if ok {
+		return n
+	}
+	f.lock.Lock()
+	defer f.lock.Unlock()
+	// check again to avoid race condition
+	n, ok = f.notifyChanges[c]
+	if ok {
+		return n
+	}
+	n = sync.NewCond(&f.lock)
+	f.notifyChanges[c] = n
+	return n
+}
+
+func (f *Filters) notify(c clusterID) {
+	f.lock.RLock()
+	defer f.lock.RUnlock()
+	if notifier, ok := f.notifyChanges[c]; ok {
+		notifier.Broadcast()
+	}
+}
+
+func (f *Filters) clearClusterPatterns(c clusterID) {
+	defer f.notify(c)
+	f.lock.Lock()
+	defer f.lock.Unlock()
+	delete(f.filters, c)
+}
+
+// copyPatternWithLock gets a copy of a cluster's filters
+func (f *Filters) copyPatternWithLock(c clusterID) *ClusterFilter {
+	filters := &ClusterFilter{}
+	if got, ok := f.filters[c]; ok {
+		filters.patterns = got.patterns.Clone()
+	} else {
+		filters.patterns = sets.New[pattern]()
+	}
+	return filters
+}
+
+// applyChanges applies the changes in the given list, atomically.
+func (f *Filters) applyChanges(c clusterID, changes []FilterChange) {
+	defer f.notify(c)
+	f.lock.Lock()
+	defer f.lock.Unlock()
+	var newPatterns sets.Set[pattern]
+	if existing, ok := f.filters[c]; ok {
+		newPatterns = existing.patterns
+	} else {
+		newPatterns = sets.New[pattern]()
+	}
+	for _, change := range changes {
+		applyChange(newPatterns, &change)
+	}
+	f.filters[c] = &ClusterFilter{patterns: newPatterns}
+}
+
+// applyChange applies a single filter change to the given set.
+func applyChange(s sets.Set[pattern], change *FilterChange) {
+	switch change.Operation {
+	case FilterChangeAdd:
+		nsPatterns := slices.Clone(change.NamespacePatterns)
+		sort.Strings(nsPatterns)
+		p := pattern{eventTypePattern: change.EventTypePattern, namespacePatterns: strings.Join(nsPatterns, " ")}
+		s.Insert(p)
+	case FilterChangeRemove:
+		nsPatterns := slices.Clone(change.NamespacePatterns)
+		sort.Strings(nsPatterns)
+		check := pattern{eventTypePattern: change.EventTypePattern, namespacePatterns: strings.Join(nsPatterns, " ")}
+		s.Delete(check)
+	case FilterChangeClear:
+		s.Clear()
 	}
 }
 
 // addPattern adds a pattern to a node's list.
-func (f *Filters) addPattern(node nodeID, namespacePatterns []string, eventTypePattern string) {
+func (f *Filters) addPattern(c clusterID, namespacePatterns []string, eventTypePattern string) {
+	defer f.notify(c)
 	f.lock.Lock()
 	defer f.lock.Unlock()
-	if _, ok := f.filters[node]; !ok {
-		f.filters[node] = &NodeFilter{}
+	if _, ok := f.filters[c]; !ok {
+		f.filters[c] = &ClusterFilter{
+			patterns: sets.New[pattern](),
+		}
 	}
 	nsPatterns := slices.Clone(namespacePatterns)
 	sort.Strings(nsPatterns)
-	f.filters[node].patterns = append(f.filters[node].patterns, pattern{eventTypePattern: eventTypePattern, namespacePatterns: nsPatterns})
+	p := pattern{eventTypePattern: eventTypePattern, namespacePatterns: strings.Join(nsPatterns, " ")}
```

Comment:
may want to sanitize `namespacePatterns` here, suggest trimming the result of `strings.Join()` or returning an error if `namespacePatterns` contains unsupported patterns like `" "` or empty string (if it is unsupported)